`@Owning` should be a type annotation

If a fix for #4932 is merged before a fix for this issue is merged, then some tests will need to be re-enabled.

`org.checkerframework.checker.mustcall.qual.Owning` is currently defined as a declaration annotation (not a type annotation):

```java
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.METHOD, ElementType.PARAMETER, ElementType.FIELD})
public @interface Owning {}
```

This ensures that `@Owning` is only written on methods, parameters, and fields.  However, it is problematic for two reasons.  First, its documentation indicates that it is about types rather than parameters.  Second, in Java 17 a declaration annotation cannot be written on the `this` type parameter.  Under Java 17, running the tests yields

```
org.checkerframework.checker.test.junit.MustCallTest > run[/home/mernst/research/types/checker-framework-fork-mernst-branch-jdk17/checker/mustcall] FAILED
    java.lang.AssertionError: 0 out of 40 expected diagnostics were found.
    1 unexpected diagnostic was found
    OwningParams.java:8: error: annotation @org.checkerframework.checker.mustcall.qual.Owning not applicable in this type context
```

--- a/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/NotOwning.java
+++ b/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/NotOwning.java
@@ -7,7 +7,8 @@ import java.lang.annotation.Target;
 
 /**
  * Annotation indicating that ownership should not be transferred to the annotated parameter, field,
- * or (when this is written on a method) return value, for the purposes of Must Call checking.
+ * or method's call sites, for the purposes of Must Call checking. For a full description of the
+ * semantics, see the documentation of {@link Owning}.
  *
  * <p>Parameters and fields are treated as if they have this annotation by default unless they have
  * {@link Owning}.

--- a/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/NotOwning.java
+++ b/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/NotOwning.java
@@ -7,7 +7,7 @@ import java.lang.annotation.Target;
 
 /**
  * Annotation indicating that ownership should not be transferred to the annotated parameter, field,
- * or (when this is written on a method) return type, for the purposes of Must Call checking.
+ * or (when this is written on a method) return value, for the purposes of Must Call checking.
  *
  * <p>Parameters and fields are treated as if they have this annotation by default unless they have
  * {@link Owning}.

--- a/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/Owning.java
+++ b/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/Owning.java
@@ -6,21 +6,26 @@ import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
 /**
- * Annotation indicating that ownership should be transferred to the annotated parameter, field, or
- * (when written on a method) return value, for the purposes of Must Call checking. Static fields
- * cannot be owning.
+ * Annotation indicating that ownership should be transferred to the annotated element for the
+ * purposes of Must Call checking. When written on a parameter, the annotation indicates that Must
+ * Call checking should be performed in the body of the method, not at call sites. When written on a
+ * method, the annotation indicates that return expressions do not need to be checked in the method
+ * body, but at call sites. When written on a field, the annotation indicates that fulfilling the
+ * must-call obligations of an instance of the class in which the field is declared also results in
+ * the annotated field's must-call obligations being satisfied. Static fields cannot be owning.
  *
  * <p>This annotation is a declaration annotation rather than a type annotation, because it does not
  * logically belong to any type hierarchy. Logically, it is a directive to the Resource Leak Checker
- * that informs it whether checking that a value's must-call obligations have been satisfied is
- * necessary, or not. In that way, it can be viewed as an annotation expressing an aliasing
- * relationship: passing a object with a non-empty must-call obligation to a method with an owning
- * parameter resolves that object's must-call obligation, because the ownership annotation expresses
- * that the object at the call site and the parameter in the method's body are aliases, and so
- * checking only one of the two is required.
+ * that informs it whether it is necessary to check that a value's must-call obligations have been
+ * satisfied. In that way, it can be viewed as an annotation expressing an aliasing relationship:
+ * passing a object with a non-empty must-call obligation to a method with an owning parameter
+ * resolves that object's must-call obligation, because the ownership annotation expresses that the
+ * object at the call site and the parameter in the method's body are aliases, and so checking only
+ * one of the two is required.
  *
- * <p>Method return values are treated as if they have this annotation by default unless their
- * method is annotated as {@link NotOwning}.
+ * <p>Methods are treated as if they have this annotation by default unless their method is
+ * annotated as {@link NotOwning}. Parameters and fields are treated as {@link NotOwning} by
+ * default.
  *
  * <p>When the -AnoLightweightOwnership command-line argument is passed to the checker, this
  * annotation and {@link NotOwning} are ignored.

--- a/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/Owning.java
+++ b/checker-qual/src/main/java/org/checkerframework/checker/mustcall/qual/Owning.java
@@ -7,11 +7,20 @@ import java.lang.annotation.Target;
 
 /**
  * Annotation indicating that ownership should be transferred to the annotated parameter, field, or
- * (when written on a method) return type, for the purposes of Must Call checking. Static fields
+ * (when written on a method) return value, for the purposes of Must Call checking. Static fields
  * cannot be owning.
  *
- * <p>Method return types are treated as if they have this annotation by default unless their method
- * is annotated as {@link NotOwning}.
+ * <p>This annotation is a declaration annotation rather than a type annotation, because it does not
+ * logically belong to any type hierarchy. Logically, it is a directive to the Resource Leak Checker
+ * that informs it whether checking that a value's must-call obligations have been satisfied is
+ * necessary, or not. In that way, it can be viewed as an annotation expressing an aliasing
+ * relationship: passing a object with a non-empty must-call obligation to a method with an owning
+ * parameter resolves that object's must-call obligation, because the ownership annotation expresses
+ * that the object at the call site and the parameter in the method's body are aliases, and so
+ * checking only one of the two is required.
+ *
+ * <p>Method return values are treated as if they have this annotation by default unless their
+ * method is annotated as {@link NotOwning}.
  *
  * <p>When the -AnoLightweightOwnership command-line argument is passed to the checker, this
  * annotation and {@link NotOwning} are ignored.

--- a/checker/tests/mustcall/OwningParams.java
+++ b/checker/tests/mustcall/OwningParams.java
@@ -1,21 +1,13 @@
-// @above-java11-skip-test until this is fixed: https://tinyurl.com/cfissue/4934
-
-// Tests that parameters (including receiver parameters) marked as @Owning are still checked.
+// Tests that parameters marked as @Owning are still checked.
 
 import org.checkerframework.checker.mustcall.qual.*;
 
 class OwningParams {
   static void o1(@Owning OwningParams o) {}
 
-  void o2(@Owning OwningParams this) {}
-
   void test(@Owning @MustCall({"a"}) OwningParams o, @Owning OwningParams p) {
     // :: error: argument
     o1(o);
-    // TODO: this error doesn't show up! See MustCallVisitor#skipReceiverSubtypeCheck
-    //  error: method.invocation
-    o.o2();
     o1(p);
-    p.o2();
   }
 }

